kubeflow/trainer/backends/docker/backend.py

```diff
@@ -0,0 +1,536 @@
+# Copyright 2025 The Kubeflow Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+DockerBackend
+-------------
+
+Local execution backend for `CustomTrainer` jobs using Docker containers.
+
+Key behaviors:
+- Multi-node jobs: one container per node connected via a per-job Docker network.
+- Entry script generation: we serialize the user's training function to a small
+  Python file and invoke it inside the container using `torchrun` (preferred) or
+  `python` as a fallback.
+- Runtimes: we use `config/local_runtimes` to define runtime images and
+  characteristics (e.g., torch). Defaults to `torch-distributed` if no runtime
+  is provided.
+- Image pulling: controlled via `pull_policy` and performed automatically if
+  needed.
+- Logs and lifecycle: streaming logs and deletion semantics similar to the
+  Kubernetes backend, but tailored to Docker.
+"""
+
+from __future__ import annotations
+
+from collections.abc import Iterator
+from dataclasses import dataclass
+from datetime import datetime
+import logging
+import os
+from pathlib import Path
+import random
+import shutil
+import string
+import uuid
+
+try:
+    import docker  # type: ignore
```

Comment:
How difficult would it be to support Podman as well: https://pypi.org/project/podman/

Should it be a different backend or could this be a "container" backend that could work for both Docker and Podman depending on what's installed / configured?

Reply:
I would agree with @astefanutti to unify all container-based functionality under single backend.
From the end-user point of view it doesn't matter whether they use Docker, Podman, or Container backend to run local containers.